File: 2020/day4.py
"""
byr (Birth Year)
iyr (Issue Year)
eyr (Expiration Year)
hgt (Height)
hcl (Hair Color)
ecl (Eye Color)
pid (Passport ID)
cid (Country ID)
"""
import re

with open("day4.txt") as f:
    data = f.read()

lines = data.split("\n\n")

conditions = set("byr iyr eyr hgt hcl ecl pid".split(" "))


def valid(parts):
    keys = {k.split(":")[0] for k in parts}
    if all(c in keys for c in conditions):
        return True
    else:
        return False


v = 0
for line in lines:
    parts = re.split(r"\s", line)
    if valid(parts):
        v += 1

print(v)

"""
byr (Birth Year) - four digits; at least 1920 and at most 2002.
iyr (Issue Year) - four digits; at least 2010 and at most 2020.
eyr (Expiration Year) - four digits; at least 2020 and at most 2030.
hgt (Height) - a number followed by either cm or in:
    If cm, the number must be at least 150 and at most 193.
    If in, the number must be at least 59 and at most 76.
hcl (Hair Color) - a # followed by exactly six characters 0-9 or a-f.
ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth.
pid (Passport ID) - a nine-digit number, including leading zeroes.
cid (Country ID) - ignored, missing or not
"""


def apply_condition(key, value):
    if key == "byr":
        res = int(value) >= 1920 and int(value) <= 2002
    if key == "iyr":
        res = int(value) >= 2010 and int(value) <= 2020
    if key == "eyr":
        res = int(value) >= 2020 and int(value) <= 2030
    if key == "hgt":
        num, in_or_cm = re.findall(r"^(\d+)(cm|in)$", value)[0]
        if in_or_cm == "in":
            res = int(num) >= 59 and int(num) <= 76
        elif in_or_cm == "cm":
            res = int(num) >= 150 and int(num) <= 193
        else:
            res = False
    if key == "hcl":
        res = bool(re.match(r"^#[0-9a-f]{6}$", value))
    if key == "ecl":
        res = value in {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"}
    if key == "pid":
        res = bool(re.match(r"^[0-9]{9}$", value))
    if key == "cid":
        res = True
    return res


def valid(parts):
    conds = {k.split(":")[0]: k.split(":")[1] for k in parts if k}
    if not all(c in conds for c in conditions):
        return False
    try:
        res = [apply_condition(key, value) for key, value in conds.items()]
        return all(res)

    except:
        return False


v = 0
for line in lines:
    parts = re.split(r"\s", line)
    if valid(parts):
        v += 1

print(v)

File: 2020/day4.exs
defmodule Day4 do
  @re ~r/((\w{3}:[\w#]+) ?\n?)+/m
  @must_be ~w(byr iyr eyr hgt hcl ecl pid)
  @valid_eye MapSet.new(~w(amb blu brn gry grn hzl oth))
  @four_digit_re ~r<^\d{4}$>
  @hcl_re ~r/^#[0-9a-f]{6}$/
  @pid_re ~r/^\d{9}$/

  def run(part) do
    contents = File.read!("day4.txt")

    @re
    |> Regex.scan(contents)
    |> filter(part)
    |> length
  end

  def filter(parsed, :one) do
    Enum.filter(parsed, &contains_all?/1)
  end

  def filter(parsed, _) do
    Enum.filter(parsed, fn x -> contains_all?(x) and rules?(x) end)
  end

  def contains_all?([full_match | _]) do
    Enum.all?(@must_be, &String.contains?(full_match, &1))
  end

  def rules?([full_match | _]) do
    Regex.split(~r/\s/m, full_match, trim: true)
    |> Stream.map(&String.split(&1, ":"))
    |> Enum.all?(&valid?/1)
  end

  def valid?(["cid", _]), do: true
  def valid?(["byr", num]), do: valid_date?(num, 1920, 2002)
  def valid?(["iyr", num]), do: valid_date?(num, 2010, 2020)
  def valid?(["eyr", num]), do: valid_date?(num, 2020, 2030)
  def valid?(["hgt", num]), do: valid_height?(num)
  def valid?(["hcl", string]), do: valid_match?(@hcl_re, string)
  def valid?(["pid", num]), do: valid_match?(@pid_re, num)
  def valid?(["ecl", string]), do: MapSet.member?(@valid_eye, string)

  def valid_match?(re, string) do
    Regex.match?(re, string)
  end

  def cond_for(num, units, lower, upper) do
    without_suffix = String.trim_trailing(num, units)

    String.ends_with?(num, units) and
      valid_limits?(without_suffix, lower, upper)
  end

  def valid_height?(num) do
    cond do
      cond_for(num, "in", 59, 76) ->
        true

      cond_for(num, "cm", 150, 193) ->
        true

      true ->
        false
    end
  end

  def valid_limits?(num, lower, upper) do
    case Integer.parse(num) do
      {n, ""} -> n >= lower and n <= upper
      _ -> false
    end
  end

  def valid_date?(num, lower, upper) do
    Regex.match?(@four_digit_re, num) and valid_limits?(num, lower, upper)
  end
end

Day4.run(:one)
|> IO.inspect()

Day4.run(:two)
|> IO.inspect()
